File: kong/helm-chart/lua-scripts/custom-jwt-auth.lua
local http = require "resty.http"
local cjson = require "cjson"

-- Global cache for public keys and configuration
local JWKS_CACHE = {}
local CACHE_EXPIRY = 0
local CACHE_TTL = 3600 -- 1 hour in seconds

-- Keycloak configuration (will be defined by Helm template before this script)
-- Variables KEYCLOAK_BASE_URL, KEYCLOAK_REALM, EXPECTED_ISSUER, EXPECTED_AUDIENCE
-- are expected to be defined before this script is loaded
local JWKS_URL = KEYCLOAK_BASE_URL .. "/realms/" .. KEYCLOAK_REALM .. "/protocol/openid-connect/certs"

-- Utility function to log with request ID
local function log_info(msg)
  kong.log.info("[custom-jwt] " .. msg)
end

local function log_debug(msg)
  kong.log.debug("[custom-jwt] " .. msg)
end

local function log_error(msg)
  kong.log.err("[custom-jwt] " .. msg)
end

-- Function to fetch JWKS from Keycloak
local function fetch_jwks()
  log_debug("Fetching JWKS from: " .. JWKS_URL)
  
  local httpc = http.new()
  httpc:set_timeout(10000) -- 10 seconds
  
  local res, err = httpc:request_uri(JWKS_URL, {
    method = "GET",
    headers = {
      ["Accept"] = "application/json",
      ["User-Agent"] = "Kong-Custom-JWT/1.0"
    },
    ssl_verify = false
  })
  
  if not res then
    log_error("Failed to fetch JWKS: " .. (err or "unknown error"))
    return nil
  end
  
  if res.status ~= 200 then
    log_error("JWKS endpoint returned status " .. res.status)
    return nil
  end
  
  local success, jwks = pcall(cjson.decode, res.body)
  if not success then
    log_error("Failed to parse JWKS JSON: " .. (jwks or "invalid JSON"))
    return nil
  end
  
  log_info("Successfully fetched JWKS with " .. #(jwks.keys or {}) .. " keys")
  return jwks
end

-- Function to get public keys with caching
local function get_public_keys()
  local current_time = ngx.time()
  
  -- Return cached keys if still valid
  if current_time < CACHE_EXPIRY and next(JWKS_CACHE) ~= nil then
    log_debug("Using cached JWKS (expires in " .. (CACHE_EXPIRY - current_time) .. " seconds)")
    return JWKS_CACHE
  end
  
  -- Fetch fresh JWKS
  log_debug("Cache expired or empty, fetching fresh JWKS")
  local jwks = fetch_jwks()
  
  if not jwks or not jwks.keys then
    log_error("Failed to fetch valid JWKS")
    -- Return cached keys if available, even if expired
    if next(JWKS_CACHE) ~= nil then
      log_info("Returning expired cached JWKS as fallback")
      return JWKS_CACHE
    end
    return nil
  end
  
  -- Process and cache the keys
  local key_cache = {}
  for _, key in ipairs(jwks.keys) do
    if key.kid and key.kty == "RSA" and key.use == "sig" then
      -- Convert JWK to PEM format for resty.jwt
      local n = key.n
      local e = key.e
      
      if n and e then
        key_cache[key.kid] = {
          kty = key.kty,
          n = n,
          e = e,
          alg = key.alg or "RS256"
        }
        log_debug("Cached key with kid: " .. key.kid)
      end
    end
  end
  
  JWKS_CACHE = key_cache
  CACHE_EXPIRY = current_time + CACHE_TTL
  
  log_info("Cached " .. #jwks.keys .. " keys (expires at " .. CACHE_EXPIRY .. ")")
  return JWKS_CACHE
end

-- Helper function to base64url decode
local function base64url_decode(str)
  -- Add padding if needed
  local padding = 4 - (#str % 4)
  if padding ~= 4 then
    str = str .. string.rep("=", padding)
  end
  -- Replace URL-safe characters
  str = str:gsub("-", "+"):gsub("_", "/")
  return ngx.decode_base64(str)
end

-- Helper function to parse JWT without verification
local function parse_jwt(token)
  local parts = {}
  for part in token:gmatch("[^%.]+") do
    table.insert(parts, part)
  end
  
  if #parts ~= 3 then
    return nil, "Invalid JWT format - expected 3 parts"
  end
  
  -- Decode header
  local header_json = base64url_decode(parts[1])
  if not header_json then
    return nil, "Failed to decode JWT header"
  end
  
  local success, header = pcall(cjson.decode, header_json)
  if not success then
    return nil, "Failed to parse JWT header JSON"
  end
  
  -- Decode payload
  local payload_json = base64url_decode(parts[2])
  if not payload_json then
    return nil, "Failed to decode JWT payload"
  end
  
  local success, payload = pcall(cjson.decode, payload_json)
  if not success then
    return nil, "Failed to parse JWT payload JSON"
  end
  
  return {
    header = header,
    payload = payload,
    signature = parts[3],
    signed_content = parts[1] .. "." .. parts[2]
  }, nil
end

-- Function to verify JWT token
local function verify_jwt_token(token)
  log_debug("Starting JWT verification")
  
  -- Remove Bearer prefix if present
  if token:sub(1, 7):lower() == "bearer " then
    token = token:sub(8)
    log_debug("Removed Bearer prefix from token")
  end
  
  -- Parse JWT token
  local jwt_obj, parse_error = parse_jwt(token)
  if not jwt_obj then
    log_error("Failed to parse JWT: " .. (parse_error or "unknown error"))
    return false, "Invalid JWT format"
  end
  
  local kid = jwt_obj.header.kid
  local alg = jwt_obj.header.alg
  
  if not kid then
    log_error("No kid found in JWT header")
    return false, "Missing key ID in JWT header"
  end
  
  if alg ~= "RS256" then
    log_error("Unsupported algorithm: " .. (alg or "unknown"))
    return false, "Unsupported JWT algorithm"
  end
  
  log_debug("JWT kid: " .. kid .. ", algorithm: " .. alg)
  
  -- Get public keys
  local public_keys = get_public_keys()
  if not public_keys then
    log_error("No public keys available")
    return false, "Unable to fetch public keys"
  end
  
  -- Find the public key for this kid
  local key_data = public_keys[kid]
  if not key_data then
    log_error("Key ID " .. kid .. " not found in JWKS")
    local key_ids = {}
    for k, _ in pairs(public_keys) do
      table.insert(key_ids, k)
    end
    log_debug("Available key IDs: " .. table.concat(key_ids, ", "))
    return false, "Key not found"
  end
  
  -- For now, skip signature verification and just validate claims
  -- TODO: Implement RSA signature verification using resty.rsa
  log_debug("Skipping signature verification (would need resty.rsa implementation)")
  
  -- Verify standard claims
  local payload = jwt_obj.payload
  if not payload then
    log_error("No payload in JWT")
    return false, "Invalid JWT payload"
  end
  
  -- Check expiration
  local now = ngx.time()
  if payload.exp and payload.exp < now then
    log_error("JWT token expired")
    return false, "Token expired"
  end
  
  -- Check issuer
  if payload.iss ~= EXPECTED_ISSUER then
    log_error("Invalid issuer: " .. (payload.iss or "none") .. " (expected: " .. EXPECTED_ISSUER .. ")")
    return false, "Invalid issuer"
  end
  
  -- Check audience
  local valid_audience = false
  if type(payload.aud) == "string" then
    valid_audience = payload.aud == EXPECTED_AUDIENCE
  elseif type(payload.aud) == "table" then
    for _, aud in ipairs(payload.aud) do
      if aud == EXPECTED_AUDIENCE then
        valid_audience = true
        break
      end
    end
  end
  
  if not valid_audience then
    log_error("Invalid audience")
    return false, "Invalid audience"
  end
  
  log_info("JWT verification successful for subject: " .. (payload.sub or "unknown"))
  
  -- Add user info to headers for downstream services
  kong.service.request.set_header("X-User-ID", payload.sub or "")
  kong.service.request.set_header("X-Client-ID", payload.client_id or "")
  kong.service.request.set_header("X-Username", payload.preferred_username or "")
  
  return true, "Token valid"
end

-- Main execution function
local function main()
  -- Get Authorization header
  local auth_header = kong.request.get_header("Authorization")
  
  if not auth_header then
    log_info("No Authorization header found")
    return kong.response.exit(401, {
      error = "Authorization Required",
      message = "Missing Authorization header"
    })
  end
  
  -- Verify the JWT token
  local valid, error_message = verify_jwt_token(auth_header)
  
  if not valid then
    log_info("Authentication failed: " .. error_message)
    return kong.response.exit(401, {
      error = "Authentication Failed", 
      message = error_message
    })
  end
  
  log_debug("Authentication successful, proceeding to upstream")
end

-- Execute main function
main()

File: kong/helm-chart/python-plugins/custom-jwt-auth.py
"""
Custom JWT Authentication Plugin for Kong Gateway
Python PDK implementation
"""

import kong_pdk
import json
import time
import base64
import requests
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
import jwt as pyjwt

# Plugin schema
Schema = (
    {
        "keycloak_base_url": {
            "type": "string", 
            "required": True,
            "description": "Keycloak base URL (e.g., https://keycloak.example.com)"
        }
    },
    {
        "keycloak_realm": {
            "type": "string",
            "required": True, 
            "description": "Keycloak realm name"
        }
    },
    {
        "expected_issuer": {
            "type": "string",
            "required": True,
            "description": "Expected JWT issuer"
        }
    },
    {
        "expected_audience": {
            "type": "string", 
            "required": True,
            "description": "Expected JWT audience"
        }
    },
    {
        "cache_ttl": {
            "type": "number",
            "default": 3600,
            "description": "JWKS cache TTL in seconds"
        }
    },
    {
        "ssl_verify": {
            "type": "boolean",
            "default": False,
            "description": "Whether to verify SSL certificates"
        }
    }
)

# Plugin metadata
version = "1.0.0"
priority = 1000


class Plugin:
    """
    Custom JWT Authentication Plugin using Kong Python PDK
    """
    
    def __init__(self, config):
        self.config = config
        self.jwks_cache = {}
        self.cache_expiry = 0
        self.cache_ttl = config.get("cache_ttl", 3600)  # Use config value or default to 1 hour
    
    def access(self, kong):
        """
        Main access phase implementation
        Validates JWT tokens using Keycloak JWKS
        """
        try:
            # Get plugin configuration from instance variable
            keycloak_base_url = self.config.get("keycloak_base_url")
            keycloak_realm = self.config.get("keycloak_realm")
            expected_issuer = self.config.get("expected_issuer")
            expected_audience = self.config.get("expected_audience")
            
            if not all([keycloak_base_url, keycloak_realm, expected_issuer, expected_audience]):
                kong.log.err("[custom-jwt] Missing required configuration")
                return kong.response.exit(500, {"error": "Plugin configuration error"})
            
            # Get Authorization header
            auth_header = kong.request.get_header("Authorization")
            
            # Handle case where header might be returned as tuple
            if isinstance(auth_header, tuple):
                auth_header = auth_header[0] if auth_header else None
            
            if not auth_header:
                kong.log.info("[custom-jwt] No Authorization header found")
                return kong.response.exit(401, {
                    "error": "Authorization Required",
                    "message": "Missing Authorization header"
                })
            
            # Extract token from Bearer authorization
            if not auth_header.startswith("Bearer "):
                kong.log.info("[custom-jwt] Invalid Authorization header format")
                return kong.response.exit(401, {
                    "error": "Invalid Authorization Format",
                    "message": "Authorization header must start with 'Bearer '"
                })
            
            jwt_token = auth_header[7:]  # Remove "Bearer " prefix
            
            # Validate JWT token
            is_valid, payload, error_msg = self._validate_jwt(
                jwt_token, 
                keycloak_base_url, 
                keycloak_realm,
                expected_issuer,
                expected_audience,
                self.config.get("ssl_verify", False)
            )
            
            if not is_valid:
                kong.log.warn(f"[custom-jwt] JWT validation failed: {error_msg}")
                return kong.response.exit(401, {
                    "error": "Invalid Token",
                    "message": error_msg
                })
            
            # Add user info to headers for upstream services
            if payload:
                # Extract user information from JWT claims
                user_id = payload.get("sub", "")
                username = payload.get("preferred_username", "")
                email = payload.get("email", "")
                roles = payload.get("realm_access", {}).get("roles", [])
                
                # Set headers for upstream
                kong.service.request.set_header("X-User-ID", user_id)
                kong.service.request.set_header("X-Username", username)
                kong.service.request.set_header("X-User-Email", email)
                kong.service.request.set_header("X-User-Roles", ",".join(roles))
                
                kong.log.info(f"[custom-jwt] Authentication successful for user: {username}")
            
        except Exception as e:
            kong.log.err(f"[custom-jwt] Unexpected error: {str(e)}")
            return kong.response.exit(500, {
                "error": "Internal Server Error",
                "message": "Authentication service error"
            })
    
    def _validate_jwt(self, token, keycloak_base_url, realm, expected_issuer, expected_audience, ssl_verify):
        """
        Validate JWT token using Keycloak's public key from JWKS endpoint
        Returns: (is_valid, payload, error_message)
        """
        try:
            # First, decode JWT header to get the key ID (kid)
            try:
                header = pyjwt.get_unverified_header(token)
                kid = header.get('kid')
                alg = header.get('alg', 'RS256')
                
                if alg != 'RS256':
                    return False, None, f"Unsupported algorithm: {alg}"
                
                if not kid:
                    return False, None, "Missing key ID (kid) in JWT header"
                    
            except Exception as e:
                return False, None, f"Invalid JWT header: {str(e)}"
            
            # Get public key from JWKS
            public_key, error_msg = self._get_public_key_from_jwks(
                keycloak_base_url, realm, kid, ssl_verify
            )
            
            if not public_key:
                return False, None, error_msg or "Failed to get public key"
            
            # Validate JWT signature and claims
            try:
                # Decode and verify the token
                payload = pyjwt.decode(
                    token,
                    public_key,
                    algorithms=['RS256'],
                    issuer=expected_issuer,
                    audience=expected_audience,
                    options={
                        'verify_signature': True,
                        'verify_exp': True,
                        'verify_nbf': True,
                        'verify_iat': True,
                        'verify_aud': True,
                        'verify_iss': True,
                        'require_exp': True,
                        'require_iat': True,
                        'require_nbf': False
                    }
                )
                
                return True, payload, None
                
            except pyjwt.ExpiredSignatureError:
                return False, None, "Token has expired"
            except pyjwt.InvalidTokenError as e:
                return False, None, f"Invalid token: {str(e)}"
            except pyjwt.InvalidIssuerError:
                return False, None, f"Invalid issuer: expected {expected_issuer}"
            except pyjwt.InvalidAudienceError:
                return False, None, f"Invalid audience: expected {expected_audience}"
            except pyjwt.InvalidSignatureError:
                return False, None, "Invalid token signature"
            except Exception as e:
                return False, None, f"Token validation error: {str(e)}"
            
        except Exception as e:
            return False, None, f"JWT validation error: {str(e)}"
    
    def _get_public_key_from_jwks(self, keycloak_base_url, realm, kid, ssl_verify):
        """
        Fetch public key from Keycloak JWKS endpoint
        Returns: (public_key, error_message)
        """
        try:
            # Check cache first
            current_time = time.time()
            cache_key = f"{keycloak_base_url}/{realm}"
            
            if (cache_key in self.jwks_cache and 
                current_time < self.cache_expiry):
                jwks = self.jwks_cache[cache_key]
            else:
                # Fetch JWKS from Keycloak
                jwks_url = f"{keycloak_base_url}/realms/{realm}/protocol/openid-connect/certs"
                
                try:
                    response = requests.get(jwks_url, verify=ssl_verify, timeout=10)
                    response.raise_for_status()
                    jwks = response.json()
                    
                    # Cache the JWKS
                    self.jwks_cache[cache_key] = jwks
                    self.cache_expiry = current_time + self.cache_ttl
                    
                except requests.RequestException as e:
                    return None, f"Failed to fetch JWKS from {jwks_url}: {str(e)}"
                except json.JSONDecodeError as e:
                    return None, f"Invalid JWKS response from {jwks_url}: {str(e)}"
            
            # Find the key with matching kid
            keys = jwks.get('keys', [])
            target_key = None
            
            for key in keys:
                if key.get('kid') == kid:
                    target_key = key
                    break
            
            if not target_key:
                return None, f"Key with kid '{kid}' not found in JWKS"
            
            # Validate key properties
            if target_key.get('kty') != 'RSA':
                return None, f"Unsupported key type: {target_key.get('kty')}"
            
            if target_key.get('use') not in ['sig', None]:
                return None, f"Key not for signature use: {target_key.get('use')}"
            
            # Extract RSA key components
            try:
                n = target_key.get('n')  # Modulus
                e = target_key.get('e')  # Exponent
                
                if not n or not e:
                    return None, "Missing RSA key components (n or e)"
                
                # Decode base64url values
                n_bytes = self._base64url_decode(n)
                e_bytes = self._base64url_decode(e)
                
                # Convert to integers
                n_int = int.from_bytes(n_bytes, byteorder='big')
                e_int = int.from_bytes(e_bytes, byteorder='big')
                
                # Create RSA public key
                public_numbers = rsa.RSAPublicNumbers(e_int, n_int)
                public_key = public_numbers.public_key()
                
                return public_key, None
                
            except Exception as e:
                return None, f"Failed to construct RSA public key: {str(e)}"
            
        except Exception as e:
            return None, f"Error getting public key from JWKS: {str(e)}"
    
    def _base64url_decode(self, data):
        """
        Decode base64url encoded data
        """
        # Add padding if needed
        padding = 4 - len(data) % 4
        if padding != 4:
            data += '=' * padding
        
        return base64.urlsafe_b64decode(data)
    
